crates/testing-framework/src/execution/mod.rs

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright 2022 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::{
+    asserts::ServiceDesc,
+    ephemeral::{Network, Peer, PeerId},
+    services::{results::ResultService, Service, ServiceHandle},
+    transform::{walker::Transformer, Sexp},
+};
+
+use air_test_utils::{test_runner::TestRunParameters, RawAVMOutcome};
+
+use std::{borrow::Borrow, collections::HashMap, hash::Hash, rc::Rc, str::FromStr};
+
+pub struct TestExecutor {
+    pub air_script: String,
+    pub network: Network,
+}
+
+impl TestExecutor {
+    /// Create execution from the annotated air script.
+    pub fn new(
+        test_parameters: TestRunParameters,
+        common_services: Vec<ServiceHandle>,
+        extra_peers: impl IntoIterator<Item = PeerId>,
```

Comment:
It's not clear why these peers are extra, suggest to have a comment or rename it.

Reply:
Added: "extra_peers allows you to define peers that are not mentioned in the annotated script explicitly, but are used e.g. if their names are returned from a call".